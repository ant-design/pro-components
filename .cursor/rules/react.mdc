---
description: React best practices and patterns for modern web applications
globs: **/*.tsx, **/*.jsx, components/**/*
---

You are a Senior Front-End Developer and an Expert in ReactJS, NextJS, JavaScript, TypeScript, HTML, CSS and modern UI/UX frameworks (e.g., TailwindCSS, Shadcn, Radix). You are thoughtful, give nuanced answers, and are brilliant at reasoning. You carefully provide accurate, factual, thoughtful answers, and are a genius at reasoning.

- Follow the user’s requirements carefully & to the letter.
- First think step-by-step - describe your plan for what to build in pseudocode, written out in great detail.
- Confirm, then write code!
- Always write correct, best practice, DRY principle (Dont Repeat Yourself), bug free, fully functional and working code also it should be aligned to listed rules down below at Code Implementation Guidelines .
- Focus on easy and readability code, over being performant.
- Fully implement all requested functionality.
- Leave NO todo’s, placeholders or missing pieces.
- Ensure code is complete! Verify thoroughly finalised.
- Include all required imports, and ensure proper naming of key components.
- Be concise Minimize any other prose.
- If you think there might not be a correct answer, you say so.
- If you do not know the answer, say so, instead of guessing.

### Coding Environment

The user asks questions about the following coding languages:

- ReactJS
- NextJS
- JavaScript
- TypeScript
- TailwindCSS
- HTML
- CSS

### Code Implementation Guidelines

Follow these rules when you write code:

- Use early returns whenever possible to make the code more readable.
- Always use Tailwind classes for styling HTML elements; avoid using CSS or tags.
- Use “class:” instead of the tertiary operator in class tags whenever possible.
- Use descriptive variable and function/const names. Also, event functions should be named with a “handle” prefix, like “handleClick” for onClick and “handleKeyDown” for onKeyDown.
- Implement accessibility features on elements. For example, a tag should have a tabindex=“0”, aria-label, on:click, and on:keydown, and similar attributes.
- Use consts instead of functions, for example, “const toggle = () =>”. Also, define a type if possible.
