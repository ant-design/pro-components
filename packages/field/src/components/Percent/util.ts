/** 获取展示符号 */
export function getSymbolByRealValue(realValue: number) {
  if (realValue === 0) {
    return null;
  }
  if (realValue > 0) {
    return '+';
  }
  return '-';
}

/** 获取颜色 */
export function getColorByRealValue(realValue: number /** ,color: string */) {
  if (realValue === 0) {
    return '#595959';
  }
  return realValue > 0 ? '#ff4d4f' : '#52c41a';
}

/** 获取到最后展示的数字 */
export function getRealTextWithPrecision(
  realValue: number,
  precision: number = 2,
) {
  return precision >= 0 ? realValue?.toFixed(precision) : realValue;
}

/**
 * 转化为数字
 * @copy from https://github.com/toss/es-toolkit/blob/32a183828c244d675f46810935e45dfefec81a54/src/compat/util/toNumber.ts#L19
 */
export function toNumber(value: any): number {
  if (typeof value === 'symbol' || value instanceof Symbol) {
    return NaN;
  }

  return Number(value);
}
