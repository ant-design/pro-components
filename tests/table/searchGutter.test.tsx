import { cleanup, render } from '@testing-library/react';
import { ProTable } from '@xxlabs/pro-components';
import { afterAll, afterEach, beforeAll, describe, expect, it } from 'vitest';
import { waitForWaitTime } from '../util';
afterEach(() => {
  cleanup();
});

describe('BasicTable SearchGutter', () => {
  const LINE_STR_COUNT = 20;
  // Mock offsetHeight
  // @ts-expect-error
  const originOffsetHeight = Object.getOwnPropertyDescriptor(HTMLElement.prototype, 'offsetHeight').get;
  Object.defineProperty(HTMLElement.prototype, 'offsetHeight', {
    get() {
      let html = this.innerHTML;
      html = html.replace(/<[^>]*>/g, '');
      const lines = Math.ceil(html.length / LINE_STR_COUNT);
      return lines * 16;
    },
  });

  // Mock getComputedStyle
  const originGetComputedStyle = window.getComputedStyle;
  window.getComputedStyle = (ele) => {
    const style = originGetComputedStyle(ele);
    return style;
  };

  beforeAll(() => {
    process.env.NODE_ENV = 'TEST';
  });

  afterAll(() => {
    Object.defineProperty(HTMLElement.prototype, 'offsetHeight', {
      get: originOffsetHeight,
    });
    window.getComputedStyle = originGetComputedStyle;
  });

  it('🎏 ProTable support searchGutter', async () => {
    const html = render(
      <ProTable
        columns={[
          {
            dataIndex: 'money',
            valueType: 'money',
          },
        ]}
        dataSource={[]}
        options={{
          fullScreen: false,
          reload: false,
          setting: false,
        }}
        rowKey="key"
        search={{
          searchGutter: [16, 24],
        }}
        size="small"
      />,
    );
    await waitForWaitTime(1200);
    const ele = html.baseElement.querySelector<HTMLDivElement>('.ant-form');
    expect(ele).toMatchSnapshot();
  });

  it('🎏 ProTable searchGutter default is [24 0]', async () => {
    const html = render(
      <ProTable
        columns={[
          {
            dataIndex: 'money',
            valueType: 'money',
          },
        ]}
        dataSource={[]}
        options={{
          fullScreen: false,
          reload: false,
          setting: false,
        }}
        rowKey="key"
        search={{
          searchGutter: 12,
        }}
        size="small"
      />,
    );
    await waitForWaitTime(1200);
    const ele = html.baseElement.querySelector<HTMLDivElement>('.ant-col');
    expect(ele?.style.paddingLeft).toBe('6px');
  });
});
