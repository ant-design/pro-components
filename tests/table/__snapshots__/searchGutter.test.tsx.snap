// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BasicTable SearchGutter > üéè ProTable support searchGutter 1`] = `
<form
  autocomplete="off"
  class="ant-form ant-form-horizontal ant-pro-query-filter ant-pro-form"
>
  <input
    style="display: none;"
    type="text"
  />
  <div
    class="ant-row ant-row-start ant-pro-query-filter-row"
    style="margin-left: -8px; margin-right: -8px; row-gap: 24px;"
  >
    <div
      class="ant-col ant-col-8 ant-pro-query-filter-row-split"
      style="padding-left: 8px; padding-right: 8px;"
    >
      <div
        class="ant-form-item"
        style="flex-wrap: nowrap;"
      >
        <div
          class="ant-row ant-form-item-row"
        >
          <div
            class="ant-col ant-form-item-control"
            style="max-width: calc(100% - 80px);"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-input-number ant-input-number-in-form-item ant-input-number-outlined"
                  style="width: 100%;"
                >
                  <div
                    class="ant-input-number-handler-wrap"
                  >
                    <span
                      aria-disabled="false"
                      aria-label="Increase Value"
                      class="ant-input-number-handler ant-input-number-handler-up"
                      role="button"
                      unselectable="on"
                    >
                      <span
                        aria-label="up"
                        class="anticon anticon-up ant-input-number-handler-up-inner"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="up"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                          />
                        </svg>
                      </span>
                    </span>
                    <span
                      aria-disabled="false"
                      aria-label="Decrease Value"
                      class="ant-input-number-handler ant-input-number-handler-down"
                      role="button"
                      unselectable="on"
                    >
                      <span
                        aria-label="down"
                        class="anticon anticon-down ant-input-number-handler-down-inner"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="down"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                          />
                        </svg>
                      </span>
                    </span>
                  </div>
                  <div
                    class="ant-input-number-input-wrap"
                  >
                    <input
                      autocomplete="off"
                      class="ant-input-number-input"
                      id="money"
                      placeholder="ËØ∑ËæìÂÖ•"
                      role="spinbutton"
                      step="1"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-8 ant-col-offset-8"
      style="padding-left: 8px; padding-right: 8px; text-align: end;"
    >
      <div
        class="ant-form-item ant-pro-query-filter-actions"
      >
        <div
          class="ant-row ant-form-item-row"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class="ant-form-item-no-colon"
              title=" "
            >
               
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-space ant-space-horizontal ant-space-align-center"
                  style="column-gap: 16px; row-gap: 16px;"
                >
                  <div
                    class="ant-space-item"
                  >
                    <div
                      style="display: flex; gap: 8px; align-items: center;"
                    >
                      <button
                        class="ant-btn ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
                        type="button"
                      >
                        <span>
                          Èáç ÁΩÆ
                        </span>
                      </button>
                      <button
                        class="ant-btn ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
                        type="button"
                      >
                        <span>
                          Êü• ËØ¢
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;
